### DISCUSSION TITLE
using library to create esp32 to esp32 over bluetooth

### DISCUSSION THREAD

**Discussion by griffinmt** [pschatzmann/arduino-audio-tools#421] [Q&A] [answered]
I am a bit of a rookie to all this, but I want to create two identical unit cor communicating within the house. Each would have an esp32-Wroom,an IMNP441 microphone, a PAM8403 audio amp with volume control and 8 ohm speaker and a push button switch to enable talking or just receive.
Was looking at streams-i2s-a2dp.ino  and  streams-a2dp-serial.ino  and wondered if the was a way to combine them to a single program to do what I want.
Can anybody please help me get started with this. I have all the parts.

TIA
Martyn

**Comment by pschatzmann** [comment, marked as answer]
A2DP is not suited for this purpose, but you can use any other network protocol instead. I suggest that you have a look at the [communicagtions-examples](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/examples-communication).

Streaming audio is quite challanging and I would not recommend this as your first project!

**Reply by griffinmt** [reply]
But I have all the parts and I dont need any high quality sound, just so I hear when my wife yells for me to come upstairs. And **where do I define what pins I am** using to read the inmp441 and write to the audio amp. I was also going to use a normally open push button for talking, otherwise the microphone remains silent.

**Comment by pschatzmann** [comment]
I suggest to use the default pins but you can change them as well. 
Just read the Wiki!

**Comment by griffinmt** [comment]
There is a lot of good stuff on your wiki. But, contrary to your earlier
suggestion, I am still looking for something that will take sound from an
INMP441 and transmit it to a receiver over bluetooth. And in the same app,
have it receive over blue tooth and feed the sound to a PAM8403 AUDIO AMP.
A single channel (left) is good enough.
Martyn

**Comment by griffinmt** [comment]
PS: I am using the wifi mac address as a unique bluetooth name.

**Comment by pschatzmann** [comment]
My answer is still the same: Bluetooth is not the right way and you need to dig into other protocols. If you don't know anthing about communication protocols you need to do quite some learning first!
You would stream the input data via a communication protocol to an output device. 
Try first to have each step working and tested first before you put the different pieces together as suggested in https://github.com/pschatzmann/arduino-audio-tools/wiki/It's-not-working

**Comment by griffinmt** [comment]
I will take your advice and consider using UDP. Was looking at the
communication examples and example-udp-send.ino with
example-udp-receive.ino and also the communication.h file.
On the send example i need to get sound from my i2s mike instead of sine
wave. That will be my first attempt.
Again, any and all help would be greatly  appreciated.
TIA
Martyn

**Comment by griffinmt** [comment]
God, i wish i could write and understand a fraction of what you do!!🙃

**Comment by pschatzmann** [comment]
The IMNP441 is using I2S so you would just replace the GeneratedSoundStream with an I2SStream as described in https://github.com/pschatzmann/arduino-audio-tools/wiki/External-ADC

**Comment by griffinmt** [comment]
I hope this is close to right!!

/**
 * @file example-udp-send.ino
 * @author Phil Schatzmann
 * @brief Sending audio over udp
 * @version 0.1
 * @date 2022-03-09
 *
 * @copyright Copyright (c) 2022
 *
 * ATTENTION: DRAFT - not tested yet
 */

#include "AudioTools.h"
#include "AudioLibs/Communication.h"

// Connections to INMP441 I2S microphone
#define I2S_WS 35 // purple
#define I2S_SD 33 // blue
#define I2S_SCK 32 // grey

uint16_t sample_rate = 44100;
uint8_t channels = 2;  // The stream will have 2 channels

I2SStream i2sStream; // Access I2S as stream

UDPStream udp;
Throttle throttle;
IPAddress udpAddress(192, 168, 0, 255);
const int udpPort = 7000;
StreamCopy copier(udp, i2sStream);  // copies sound
const char *ssid = "GRIFFIN";
const char *password = "roadkill";


void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

    auto cfg = i2sStream.defaultConfig(RX_MODE);
    cfg.i2s_format = I2S_STD_FORMAT;
    cfg.bits_per_sample = 16;
    cfg.channels = 2;
    cfg.sample_rate = 44100;
    cfg.is_master = true;
    cfg.use_apll = false;  // try with yes

//?????????????????????????
  auto i2s_pin_config_t pin_config = {
    .bck_io_num = I2S_SCK,
    .ws_io_num = I2S_WS,
    .data_out_num = -1,
    .data_in_num = I2S_SD
  };
//?????????????????????????

    i2sStream.begin(cfg);

  // connect to WIFI
  WiFi.begin(ssid, password);
  while (WiFi.status() != WL_CONNECTED) {
      delay(500);
      Serial.print(".");
  }
  Serial.println(WiFi.macAddress());
  Serial.println(WiFi.localIP());

  // Define udp address and port
  udp.begin(udpAddress, udpPort);
  auto cfg = throttle.defaultConfig();
  cfg.channels = channels;
  cfg.sample_rate = sample_rate;
  throttle.begin(cfg);

}

void loop() {
    throttle.startDelay();
    int bytes = copier.copy();
    throttle.delayBytes(bytes);
}
Martyn T. Griffin

**Comment by griffinmt** [comment]
Ignore the previous msg, I will try this one:

#include "AudioTools.h"
#include "AudioLibs/Communication.h"

// Connections to INMP441 I2S microphone
#define I2S_WS 35 // purple
#define I2S_SD 33 // blue
#define I2S_SCK 32 // grey

uint16_t sample_rate = 44100;
uint8_t channels = 2;  // The stream will have 2 channels

I2SStream i2sStream; // Access I2S as stream

UDPStream udp;
Throttle throttle;
IPAddress udpAddress(192, 168, 0, 255);
const int udpPort = 7000;
StreamCopy copier(udp, i2sStream);  // copies sound
const char *ssid = "GRIFFIN";
const char *password = "roadkill";


void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

    auto cfg = i2sStream.defaultConfig(RX_MODE);
    cfg.i2s_format = I2S_STD_FORMAT;
    cfg.bits_per_sample = 16;
    cfg.channels = 2;
    cfg.sample_rate = 44100;
    cfg.is_master = true;
    cfg.use_apll = false;  // try with yes

  auto i2s_pin_config_t = i2sStream.defaultConfig();
    i2s_pin_config_t.pin_bck = I2S_SCK;
    i2s_pin_config_t.pin_ws = I2S_WS;
    i2s_pin_config_t.pin_data = I2S_SD;
    i2s_pin_config_t.sample_rate = 44100;
    i2s_pin_config_t.channels = 2;
    i2s_pin_config_t.bits_per_sample = 16;

    i2sStream.begin(i2s_pin_config_t);

  // connect to WIFI
  WiFi.begin(ssid, password);
  while (WiFi.status() != WL_CONNECTED) {
      delay(500);
      Serial.print(".");
  }
  Serial.println(WiFi.macAddress());
  Serial.println(WiFi.localIP());

  // Define udp address and port
  udp.begin(udpAddress, udpPort);
  auto cfg2 = throttle.defaultConfig();
  cfg2.channels = channels;
  cfg2.sample_rate = sample_rate;
  throttle.begin(cfg2);

}

void loop() {
    throttle.startDelay();
    int bytes = copier.copy();
    throttle.delayBytes(bytes);
}

And this was logged:

14:45:43.867 -> [I] I2SConfig.h : 98 - rx/tx mode: TX_MODE
14:45:43.867 -> [I] I2SConfig.h : 99 - port_no: 0
14:45:43.867 -> [I] I2SConfig.h : 100 - is_master: Master
14:45:43.867 -> [I] I2SConfig.h : 101 - sample rate: 44100
14:45:43.867 -> [I] I2SConfig.h : 102 - bits per sample: 16
14:45:43.867 -> [I] I2SConfig.h : 103 - number of channels: 2
14:45:43.867 -> [I] I2SConfig.h : 104 - i2s_format: I2S_STD_FORMAT
14:45:43.914 -> [I] I2SConfig.h : 106 - auto_clear:1
14:45:43.914 -> [I] I2SConfig.h : 113 - buffer_count:10
14:45:43.914 -> [I] I2SConfig.h : 114 - buffer_size:512
14:45:43.914 -> [I] I2SConfig.h : 118 - pin_bck: 32
14:45:43.914 -> [I] I2SConfig.h : 119 - pin_ws: 35
14:45:43.914 -> [I] Ib⸮


Martyn T. Griffin
